Add WPM Temperature rgblight animation

rsheldiii opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

I saw a post on /r/mk a while back which changed the color of their underglow based on how fast the user was typing. I figured that was in qmk_firmware somewhere, but I can't find it. So, I set up this PR to change the hue of the LEDs based off Velocikey's interpretation of the user's typing speed. The color starts at 100% green (hue=120) and as the user types marches towards red (hue = 1). 

This setup requires Velocikey to be enabled in order to allow us to query the user's speed, so all the define guards reflect that. That probably isn't the most elegant solution as the code doesn't use Velocikey outside of the speed calculation, but I'd need a little help in what direction to go if there was to be a wpm module broken out.

I realize that custom animations _just_ merged like three days ago, so if this is better left in userspace let me know!

example video: https://drive.google.com/open?id=16C8JjjhV3IcvAfywlFSFUxfnorSFLYGS

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented on quantum/velocikey.c:
```diff
@@ -24,6 +24,10 @@ void velocikey_toggle(void) {
         eeprom_update_byte(EECONFIG_VELOCIKEY, 1);
 }
 
+uint8_t get_typing_speed(void) {
```
I don't think this is needed.  

If I'm reading things correctly, the `velocikey_match_speed` function should actually handle this.

In fact, I think you want to use `get_interval_time` instead. 

@chrislewisdev  Could you double check this?

rsheldiii replied:
It's definitely not needed, I thought it was conceptually cleaner. My thinking was that `velocikey_match_speed` is a function that takes a unitless `typing_speed` and returns a value in milliseconds, and I want a function that takes a unitless `typing_speed` and returns a hue value. I'm noticing now that since I can clamp `velocikey_match_speed` to whatever values I like, it's a bit of a pointless distinction. I'll update to use `get_interval_time` instead, thanks for catching that

chrislewisdev commented:
Happy to give this a closer look over the weekend :) I'm away from my main computer at the moment but once I'm back home mid next week I'll give it a test as well to make sure the 'default' velocikey usage is not impacted at all, just in case.

> I'd need a little help in what direction to go if there was to be a wpm module broken out
Velocikey essentially is a WPM module, although it was originally done for the purpose of controlling RGB animations I've kept the core of the code sufficiently isolated so that it can be used in any other context. Admittedly it does also make the `#define`s a bit messy though :)

drashna commented:
@chrislewisdev  Not a problem, at all.  Mostly just wanted your eyes on it, to verify things.  

Also, if/when you have time, it may be useful to add documentation, either to the source files themselves or to the main docs, on how to implement velocikey support.

chrislewisdev commented on quantum/rgblight.c:
```diff
@@ -1112,3 +1124,20 @@ void rgblight_effect_alternating(animation_status_t *anim) {
+  }
+
+  LED_TYPE tmp_led;
+  sethsv(201 - MAX(typing_speed, 80), rgblight_config.sat, rgblight_config.val, &tmp_led);
```
It would be awesome if we could add the smoothing/capping functionality into the Velocikey API rather than exposing the typing speed value directly like here, since that would keep all the typing-speed related code neatly in one place (and make this effect code a little neater).

If you were interested in trying to add these capabilities into Velocikey itself, I'm happy to help out, but if you'd rather just merge this code as-is, I can always give it a go myself at a later date :)

What's most important of course is that the transitions created by the code _feel_ right to you, because that's what I found to be the hardest part of writing Velocikey in the first place!

rsheldiii replied:
Looks like the capping is already there, I can work in the smoothing though! For velocikey proper I'd want to do it a little nicer, though there's a ton of ways to go about that. I'm leaning towards making `typing_speed` a floating point value and applying some kind of decay proportional to its value but I'm open to suggestions

XScorpion2 replied:
No floats pls. This has to run on AVR and be performant. Look into using the lib8tion math library to handle scaling and capping values.

chrislewisdev commented:
> Also, if/when you have time, it may be useful to add documentation, either to the source files themselves or to the main docs, on how to implement velocikey support.

Yeah, I can definitely add some comments to the API to try make it more obvious how to use, and I'll have another look at the docs to see what I can add there.

rsheldiii commented:
My only board with underglow is at work currently, so I'll implement these changes sometime this week.

rsheldiii commented:
@drashna @chrislewisdev sorry for the necro, I rebased onto master and removed the LED smoothing for another PR, after I figure out how it should work. Does this look good?

rsheldiii commented:
ah, I need to add docs support, and it looks like the value for green has shifted. I might as well scrub via `set_rgb` anyways

drashna commented:
Not a problem! 

And for the hue, we've moved from a more accurate 0-360 range for hue to a more efficient 0-255.  This means that hue fits into an unsigned 8 bit int, which also fixes a number of edge cases too.  

A quick conversion can be done, using `255 / 360 * old_hue`.  Though, using rgb may work too.

drashna commented:
any updates on this?  

aside from the previously mentioned HUE issue, it looks like the only thing that this really needs is documentation. 

if you have the time, could you add the documentation (eg, edit `/docs/feature_rgblight.md` to include the new mode?
